### PR TITLE
Fish install script

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun 

This provides an easy command `git town install-fish-autocompletion` to install the fish shell autocompletion.
This command is pretty bare bones right now, but should be all we need. Later if there is demand we can add handling of custom Fish setups etc.

Implements #284

### CODE REVIEW

**Comment by charlierudolph** [comment]
I disagree with this. To me this isn't something `git-town` should be responsible for implementing

**Comment by kevgo** [comment]
To me `git town` is the utility command for Git Town. It's job is to help the user get started, by setting Git Town up and configuring it correctly on people's machines, and displaying help for beginners.

Installing helpers like the autocompletion is a natural part of that. The scenario here is I already use Git Town, and at some point switch to fish shell. Git Town should help me make that transition easy, by offering me a command to install the integrations with that new environment. Similar to when I switch to Bash or Zsh. 

If you still disagree with this, then please come up with a better idea  :)

**Comment by charlierudolph** [comment]
If we are going to add this, can we change the API to better facilitate having other autocompletion files that can be installed (to be implemented at a later date)

```
git town install-autocompletion fish
git town install-autocompletion bash
git town install-autocompletion zch
```

File: src/helpers/terminal_helpers.sh [hunk 1]
```diff
@@ -136,3 +136,11 @@ function run_command {
   print_command "$cmd"
   eval "$cmd" 2>&1
 }
+
+
+# Run a command, prints command and output
+function run_command_outside_git {
```
Comment (charlierudolph) [inline src/helpers/terminal_helpers.sh]: How about naming this `run_command` and rename the other functions to `run_git_command` and `print_git_command`. Throw the rename into another PR that should go through quickly.

File: src/git-town [hunk 2]
```diff
@@ -2,6 +2,14 @@
 source "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/helpers/helpers.sh" --bypass-automatic-configuration --bypass-environment-checks
 
 
+# Installs the autocompletion definition for Fish shell
+function install_fish_autocompletion {
+  run_command_outside_git "mkdir -p ~/.config/fish/completions"
+  run_command_outside_git "ln -s /usr/local/Cellar/git-town/0.5/libexec/autocomplete/git.fish ~/.config/fish/completions/git.fish"
```
Comment (charlierudolph) [inline src/git-town]: This depends on the user installing via homebrew, which is not guaranteed. Why not follow the instructions on the README?

File: src/helpers/terminal_helpers.sh (hunk 1 above)
Comment (kevgo) [inline src/helpers/terminal_helpers.sh]: Good idea

**Comment by kevgo** [comment]
> If we are going to add this, can we change the API to better facilitate having other autocompletion files that can be installed (to be implemented at a later date)
>
> git town install-autocompletion fish
> git town install-autocompletion bash
> git town install-autocompletion zch

Yes, we can think about a more general interface for this, but let's implement this separately.

I am currently thinking about an API for Git Town extensions, i.e. things that are nice and useful to have, but which don't really belong into the Git Town core. Stuff related to external things like GitHub, Bitbucket, Fish, Bash, Zsh etc. That's a bigger operation however.

File: src/git-town (hunk 2 above)
Comment (kevgo) [inline src/git-town]: Following the readme now

File: src/helpers/terminal_helpers.sh (hunk 1 above)
Comment (kevgo) [inline src/helpers/terminal_helpers.sh]: Ticket for this: #376

**Comment by kevgo** [comment]
All comments addressed

**Comment by kevgo** [comment]
No, that breaks the tests. Also, feature specs should mock as little as possible, since they are full stack.

**Comment by kevgo** [comment]
I think this PR has addressed all comments. We can implement the broader API changes separately.

File: features/support/shell_overrides/ln [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+# A test mock for "ln" (creates symlinks)
+
+echo "ln called with: $@"
```
Comment (charlierudolph) [inline features/support/shell_overrides/ln]: This is no longer needed. Also shouldn't you add shell overrides for mkdir and curl?

**Comment by charlierudolph** [comment]
I think the README should now remove the reference to install fish completion. Should we also update the description of the git town command since it now has additional functionality?

**Comment by kevgo** [comment]
All documentation that I'm aware of is already updated.

**Comment by kevgo** [comment]
I have added a mock for `curl`. I cannot add a mock for `mkdir`, since we actually need to run this command in specs to create test directories. I think this command is fast enough that we can keep it as is, and don't need to mock it.

**Comment by kevgo** [comment]
All comments addressed

File: README.md [hunk 4]
```diff
@@ -110,8 +110,14 @@ so it runs anywhere Git and Bash runs.
...
-    mkdir -p ~/.config/fish/completions/
-    curl -o ~/.config/fish/completions/git.fish http://raw.githubusercontent.com/Originate/git-town/master/autocomplete/git.fish
+    $ git town install-fish-autocompletion
+    ```
+
+    or manually
+
+    ```
+    $ mkdir -p ~/.config/fish/completions/
+    $ curl -o ~/.config/fish/completions/git.fish http://raw.githubusercontent.com/Originate/git-town/master/autocomplete/git.fish
```
Comment (charlierudolph) [inline README.md]: I don't see any reason to list the manual steps.

I also think this entire section of the README can be removed and just update the summary of git-town to be 
```
git-town - general Git Town help, view and change Git Town configuration, install autocompletion
```

Comment (kevgo) [inline README.md]: Okay, removed the manual steps. I still want to at least mention how to install the autocompletions. If I would remove this here, where would this live then in your opinion?

**Comment by kevgo** [comment]
All comments addressed. Travis build is the one relevant here, which is green.

File: README.md (hunk 4 above)
Comment (charlierudolph) [inline README.md]: It already lives in the `git-town` man page and markdown page. By updating the summary of git-town as I suggested (in the README and on the man page and markdown page) I believe that points users to it.

Comment (kevgo) [inline README.md]: I see value in having all relevant installation instructions in one place in the readme. Setting Git Town up properly is the first thing people do, and at that point they are new and haven't dug through the whole documentation or man page for each command. They most likely are checking it out to see how it feels. If Git Town is not properly set up right away, and for example lacks autocompletion, it appears as a vastly inferior tool than it is, and we unnecessarily loose users. 

Telling people somewhere in a sub-page that they can also install autocompletion isn't helpful for them, nor for us. Most people don't read anything longer than a paragraph, so the installation instructions have to be concise and straight to the point.

So I'd rather keep at least the short reference to installing autocomplete in the installation instruction in the main readme, even if that is redundant. I agree it should be minimalistic, but it is at this point. 

What you propose is more logical, but less useful in real life, for real people. It isn't idiot-proof.

Does that work for you?

Comment (charlierudolph) [inline README.md]: Okay

**Comment by kevgo** [comment]
Switched to Circle, and still green. Good to go?

**Comment by charlierudolph** [comment]
:ship:

**Pull request closed without merging**